homeassistant/components/config/automation.py

```diff
@@ -107,3 +122,129 @@ async def websocket_automation_trace(hass, connection, msg):
+    {
+        vol.Required("type"): "automation/debug/breakpoint/clear",
+        vol.Required("automation_id"): str,
+        vol.Required("node"): str,
+        vol.Optional("run_id"): str,
+    }
+)
+@websocket_api.async_response
+async def websocket_automation_breakpoint_clear(hass, connection, msg):
+    """Set breakpoint."""
+    automation_id = msg.get("automation_id")
+    node = msg.get("node")
+    run_id = msg.get("run_id")
+
+    result = breakpoint_set(hass, automation_id, run_id, node)
+
+    connection.send_result(msg["id"], result)
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "automation/debug/breakpoint/list",
+    }
+)
+@websocket_api.async_response
+async def websocket_automation_breakpoint_list(hass, connection, msg):
+    """Set breakpoint."""
+    result = breakpoint_list(hass)
+
+    connection.send_result(msg["id"], result)
+
+
+@websocket_api.async_response
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "automation/debug/breakpoint/subscribe",
+    }
+)
+async def websocket_subscribe_breakpoint_events(hass, connection, msg):
+    """Subscribe to breakpoint events."""
+
+    @callback
+    def breakpoint_hit(event):
+        """Forward events to websocket."""
+        connection.send_message(
+            websocket_api.event_message(
+                msg["id"],
+                {
+                    "automation_id": event.data["automation_id"],
+                    "run_id": event.data["run_id"],
+                    "node": event.data["node"],
+                },
+            )
+        )
+
+    unsub_event = hass.bus.async_listen(EVENT_SCRIPT_BREAKPOINT_HIT, breakpoint_hit)
+
+    @callback
+    def unsub():
+        """Unsubscribe from breakpoint events."""
+        unsub_event()
+        if EVENT_SCRIPT_BREAKPOINT_HIT not in hass.bus.async_listeners:
+            breakpoint_clear_all(hass)
+
+    connection.subscriptions[msg["id"]] = unsub
+
+    connection.send_message(websocket_api.result_message(msg["id"]))
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "automation/debug/continue",
+        vol.Required("automation_id"): str,
+        vol.Required("run_id"): str,
+    }
+)
+@websocket_api.async_response
+async def websocket_automation_debug_continue(hass, connection, msg):
+    """Resume execution of halted automation."""
+    automation_id = msg.get("automation_id")
+    run_id = msg.get("run_id")
+
+    result = debug_continue(hass, automation_id, run_id)
+
+    connection.send_result(msg["id"], result)
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "automation/debug/step",
+        vol.Required("automation_id"): str,
+        vol.Required("run_id"): str,
+    }
+)
+@websocket_api.async_response
+async def websocket_automation_debug_step(hass, connection, msg):
+    """Single step a halted automation."""
+    automation_id = msg.get("automation_id")
+    run_id = msg.get("run_id")
```

Comment:
For required items, use square brackets.
```suggestion
    automation_id = msg["automation_id"]
    run_id = msg["run_id"]
```

Reply:
Fixed in bc24777